File: packages/ts/oidc-resource-server/src/discovery.ts
import { z } from 'zod';

import { OidcDiscoveryError } from './errors';

const minimalDiscoverySchema = z.object({
	issuer: z.string().min(1),
	jwks_uri: z.string().min(1),
});

export type OidcMinimalDiscovery = z.infer<typeof minimalDiscoverySchema>;

export type FetchLike = (input: string | URL, init?: RequestInit) => Promise<Response>;

function trimOrigin(origin: string): string {
	return origin.trim().replace(/\/+$/, '');
}

export async function fetchOidcDiscovery(
	apiOrigin: string,
	projectSlug: string,
	fetchImpl: FetchLike = globalThis.fetch.bind(globalThis),
): Promise<OidcMinimalDiscovery> {
	const origin = trimOrigin(apiOrigin);
	const url = `${origin}/projects/${encodeURIComponent(projectSlug)}/.well-known/openid-configuration`;
	const res = await fetchImpl(url, { method: 'GET' });
	const text = await res.text();
	if (res.status < 200 || res.status >= 300) {
		throw new OidcDiscoveryError(`Discovery failed (${res.status}): ${text}`);
	}
	let json: unknown;
	try {
		json = text.length > 0 ? JSON.parse(text) : null;
	} catch {
		throw new OidcDiscoveryError('Discovery: invalid JSON');
	}
	const parsed = minimalDiscoverySchema.safeParse(json);
	if (!parsed.success) {
		throw new OidcDiscoveryError('Discovery: missing issuer or jwks_uri');
	}
	return parsed.data;
}

export async function fetchOpenIdConfiguration(
	discoveryUrl: string,
	fetchImpl: FetchLike = globalThis.fetch.bind(globalThis),
): Promise<OidcMinimalDiscovery> {
	const res = await fetchImpl(discoveryUrl.trim(), { method: 'GET' });
	const text = await res.text();
	if (res.status < 200 || res.status >= 300) {
		throw new OidcDiscoveryError(`Discovery failed (${res.status}): ${text}`);
	}
	let json: unknown;
	try {
		json = text.length > 0 ? JSON.parse(text) : null;
	} catch {
		throw new OidcDiscoveryError('Discovery: invalid JSON');
	}
	const parsed = minimalDiscoverySchema.safeParse(json);
	if (!parsed.success) {
		throw new OidcDiscoveryError('Discovery: missing issuer or jwks_uri');
	}
	return parsed.data;
}

File: packages/dart/oidc_resource_server/lib/src/discovery.dart
import 'dart:convert';

import 'package:http/http.dart' as http;

import 'exceptions.dart';

/// OpenID Provider metadata used to verify access tokens for a project.
class OidcDiscoveryDocument {
  OidcDiscoveryDocument({required this.issuer, required this.jwksUri});

  final String issuer;
  final String jwksUri;
}

/// Fetches `/.well-known/openid-configuration` for a project-scoped issuer.
Future<OidcDiscoveryDocument> fetchOidcDiscovery({
  required String apiOrigin,
  required String projectSlug,
  http.Client? httpClient,
}) async {
  final origin = apiOrigin.trim().replaceAll(RegExp(r'/+$'), '');
  final url = Uri.parse(
    '$origin/projects/${Uri.encodeComponent(projectSlug)}/.well-known/openid-configuration',
  );
  final client = httpClient ?? http.Client();
  try {
    final res = await client.get(url);
    final raw = res.body;
    if (res.statusCode < 200 || res.statusCode >= 300) {
      throw OidcDiscoveryException('Discovery failed (${res.statusCode}): $raw');
    }
    Object? data;
    try {
      data = raw.isEmpty ? null : jsonDecode(raw);
    } catch (_) {
      throw OidcDiscoveryException('Discovery: invalid JSON');
    }
    if (data is! Map<String, dynamic>) {
      throw OidcDiscoveryException('Discovery: expected JSON object');
    }
    final issuer = data['issuer'];
    final jwks = data['jwks_uri'];
    if (issuer is! String || jwks is! String) {
      throw OidcDiscoveryException('Discovery: missing issuer or jwks_uri');
    }
    return OidcDiscoveryDocument(issuer: issuer, jwksUri: jwks);
  } finally {
    if (httpClient == null) {
      client.close();
    }
  }
}
